imports/api/governance/server/methods.ts

```diff
@@ -33,198 +34,201 @@ function mergeObjects(object1, object2, object3) {
-                let executionEpoch = referrendumEpoch && duration && duration.Referendum ? referrendumEpoch.plus(duration.Referendum) : new BigNumber(0);
-                let expirationEpoch = executionEpoch && duration && duration.Execution ? executionEpoch.plus(duration.Execution) : new BigNumber(0);
+        let approvalPhaseTime =  new BigNumber(submittedTime).plus(duration?.Approval).toNumber();
 
-                proposalData.proposal[item].proposalEpoch = proposalEpoch ? proposalEpoch.toNumber() : 0;
-                proposalData.proposal[item].referrendumEpoch = referrendumEpoch ? referrendumEpoch.toNumber() : 0;
-                proposalData.proposal[item].executionEpoch = executionEpoch ? executionEpoch.toNumber() : 0;
-                proposalData.proposal[item].expirationEpoch = expirationEpoch ? expirationEpoch.toNumber() : 0;
+        let votingPhaseEndTime = new BigNumber(approvalPhaseTime).plus(duration.Referendum).toNumber();
+        let votingPhaseStartTime = new BigNumber(votingPhaseEndTime).minus(duration?.Approval * 5).toNumber();
+        let executionPhaseStartTime = new BigNumber(votingPhaseEndTime).toNumber();
+        let executionPhaseEndTime = new BigNumber(votingPhaseEndTime).plus(duration?.Execution).toNumber();
 
-                try {
-                    proposalData.proposal[item].upvotes = (await getGovernance.getUpvotes(item)).toNumber()
+        proposalData[item].submittedTime = submittedTime ? submittedTime : 0;
+        proposalData[item].approvalPhaseTime = approvalPhaseTime ? approvalPhaseTime : 0;
+        proposalData[item].votingPhaseStartTime = votingPhaseStartTime ? votingPhaseStartTime: 0;
+        proposalData[item].votingPhaseEndTime = votingPhaseEndTime ? votingPhaseEndTime : 0;
+        proposalData[item].executionPhaseStartTime = executionPhaseStartTime ? executionPhaseStartTime: 0;
+        proposalData[item].executionPhaseEndTime = executionPhaseEndTime ? executionPhaseEndTime: 0;
 
-                }
-                catch (error) {
-                    console.log("Error when getting Governance Upvotess" + error)
-                }
+        try {
+            proposalData[item].upvotes = (await getGovernance.getUpvotes(item)).toNumber()
 
+        }
+        catch (error) {
+            console.log("Error when getting Governance Upvotess" + error)
+        }
 
-                Object.keys(proposalData.proposal).forEach(function (element) {
-                    try {
-                        Proposals.upsert(
-                            { proposalNumber: parseFloat(proposalData.proposal[element].returnValues.proposalId) },
-                            {
-                                $set: proposalData.proposal[element],
-                            }
+    }
 
-                        )
-                    } catch (e) {
-                        console.log(e);
+        Object.keys(proposalData).forEach(function (element) {
+            try {
+                Proposals.upsert(
```

Comment:
Consider upserting with mongodb bulk operations in a loop

Reply:
Ok! I have applied those changes to the code :)